stripe/public/javascript/app.js

```diff
@@ -0,0 +1,319 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const STRIPE_PUBLISHABLE_KEY = '<YOUR STRIPE PUBLISHABLE KEY>';
+let currentUser = {};
+let customerData = {};
+
+/**
+ * Firebase auth configuration
+ */
+const firebaseUI = new firebaseui.auth.AuthUI(firebase.auth());
+const firebaseUiConfig = {
+  callbacks: {
+    signInSuccessWithAuthResult: function (authResult, redirectUrl) {
+      // User successfully signed in.
+      // Return type determines whether we continue the redirect automatically
+      // or whether we leave that to developer to handle.
+      return true;
+    },
+    uiShown: () => {
+      document.getElementById('loader').style.display = 'none';
+    },
+  },
+  signInFlow: 'popup',
+  signInSuccessUrl: '/',
+  signInOptions: [
+    firebase.auth.GoogleAuthProvider.PROVIDER_ID,
+    firebase.auth.EmailAuthProvider.PROVIDER_ID,
+  ],
+  credentialHelper: firebaseui.auth.CredentialHelper.NONE,
+  // Your terms of service url.
+  tosUrl: 'https://example.com/terms',
+  // Your privacy policy url.
+  privacyPolicyUrl: 'https://example.com/privacy',
+};
+firebase.auth().onAuthStateChanged((firebaseUser) => {
+  if (firebaseUser) {
+    currentUser = firebaseUser;
+    firebase
+      .firestore()
+      .collection('stripe_customers')
+      .doc(currentUser.uid)
+      .onSnapshot((snapshot) => {
+        if (snapshot.data()) {
+          customerData = snapshot.data();
+          startDataListeners();
+          document.getElementById('loader').style.display = 'none';
+          document.getElementById('content').style.display = 'block';
+        }
+      });
+  } else {
+    document.getElementById('content').style.display = 'none';
+    firebaseUI.start('#firebaseui-auth-container', firebaseUiConfig);
+  }
+});
+
+/**
+ * Set up Stripe Elements
+ */
+const stripe = Stripe(STRIPE_PUBLISHABLE_KEY);
+const elements = stripe.elements();
+const cardElement = elements.create('card');
+cardElement.mount('#card-element');
+cardElement.on('change', ({ error }) => {
+  const displayError = document.getElementById('error-message');
+  if (error) {
+    displayError.textContent = error.message;
+  } else {
+    displayError.textContent = '';
+  }
+});
+
+/**
+ * Set up Firestore data listeners
+ */
+function startDataListeners() {
+  /**
+   * Get all payment methods for the logged in customer
+   */
+  firebase
+    .firestore()
+    .collection('stripe_customers')
+    .doc(currentUser.uid)
+    .collection('payment_methods')
+    .onSnapshot((snapshot) => {
+      if (snapshot.empty) {
+        document.querySelector('#add-new-card').open = true;
+      }
+      document
+        .querySelectorAll('select[name=payment-method] option')
+        .forEach((el) => el.remove());
```

Comment:
@jhuleatt is there a better way of doing this? I did try the `snapshot.docChanges()`[0] approach and tried to filter for `change.type === 'added'` but that didn't behave as expected and also appended too many items.

I there a way to only get the item that was added?

[0] https://firebase.google.com/docs/firestore/query-data/listen#view_changes_between_snapshots

Reply:
The way I usually do it is something like:

```js
snapshot.forEach(doc => {
  const paymentMethod = doc.data();
  if (!paymentMethod.card) {
    return;
  }

  const optionId = `card-${doc.id}`;
  let optionElement = document.getElementById(optionId);

  // add a new option if one doesn't exist yet
  if (!optionElement) {
    optionElement = document.createElement("option");
    optionElement.id = optionId;
    document.querySelector("select[name=payment-method]").appendChild(option);
  }

  optionElement.text = `${paymentMethod.card.brand} •••• ${paymentMethod.card.last4} | Expires ${paymentMethod.card.exp_month}/${paymentMethod.card.exp_year}`;
});
```

Reply:
Brilliant, I've changed to that approach, thanks :)